### PR TITLE
Generate Debug for sys types

### LINKED ISSUE #486: #[derive(Debug)] for all types in -sys crates
This would also allow to derive debug of types containing them, and generally does not hurt.

@EPashkin @GuillaumeGomez What do you think?

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #486.

cc @EPashkin @sdroege

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
All green. :)

File: src/codegen/sys/lib_.rs [hunk 1]
```diff
@@ -389,20 +440,67 @@ fn generate_classes_structs(w: &mut Write, env: &Env, classes: &[&Class]) -> Res
...
+                    \t}}\n\
+                }}\n",
+                name = klass.c_type,
             ));
         } else {
+            let can_generate_fields_debug = can_generate_fields_debug(&klass.fields);
             try!(writeln!(
                 w,
-                "{comment}#[repr(C)]\n{comment}pub struct {name} {{",
+                "{comment}#[repr(C)]\n{debug}{comment}pub struct {name} {{",
```
Comment (EPashkin) [inline src/codegen/sys/lib_.rs]: This line seems generated struct with extra line:
```
#[repr(C)]

pub struct AtkAttribute {
    pub name: *mut c_char,
    pub value: *mut c_char,
}
```

Comment (GuillaumeGomez) [reply src/codegen/sys/lib_.rs]: Wo. Where does this backline comes from? Oo

File: src/codegen/sys/lib_.rs [hunk 2]
```diff
@@ -389,20 +440,67 @@ fn generate_classes_structs(w: &mut Write, env: &Env, classes: &[&Class]) -> Res
...
+                try!(writeln!(
+                    w,
+                    "impl ::std::fmt::Debug for {name} {{\n\
+                        \tfn fmt(&self, f: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {{\n\
+                            \t\twrite!(f, \"{name} @ {{:?}} {{{{ {fields}}}}}\",\n\
+                            \t\t       self as *const _,\n\
+                            {fields_name})\n\
+                        \t}}\n\
+                    }}\n",
+                    name = klass.c_type,
```
Comment (EPashkin) [inline src/codegen/sys/lib_.rs]: Sorry, but this unreadable, and seems repeated at minimum twice.

File: src/codegen/sys/lib_.rs [hunk 3]
```diff
@@ -456,15 +572,26 @@ fn generate_records(w: &mut Write, env: &Env, records: &[&Record]) -> Result<()>
             }
             try!(writeln!(
                 w,
-                "{}#[repr(C)]\n{0}pub struct {} {{",
+                "{}#[repr(C)]\n{}\n{0}pub struct {} {{",
```
Comment (GuillaumeGomez) [inline src/codegen/sys/lib_.rs]: @EPashkin: It comes from here actually.

File: src/codegen/sys/lib_.rs (hunk 2 above)
Comment (GuillaumeGomez) [reply src/codegen/sys/lib_.rs]: True. I'll write a function and maybe comment or create temporary variables in order to make this less horrible.

File: src/codegen/sys/lib_.rs [hunk 4]
```diff
@@ -456,15 +572,26 @@ fn generate_records(w: &mut Write, env: &Env, records: &[&Record]) -> Result<()>
...
         }
+        if lines.is_empty() || !can_generate_fields_debug(&record.fields) {
+            try!(writeln!(
+                w,
+                "impl ::std::fmt::Debug for {name} {{\n\
+                    \tfn fmt(&self, f: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {{\n\
+                        \t\twrite!(f, \"{name} @ {{:?}}\", self as *const _)\n\
+                    \t}}\n\
+                }}\n",
+                name=record.c_type));
```
Comment (EPashkin) [inline src/codegen/sys/lib_.rs]: IMHO this better extract as function too and maybe use raw string for format

**Comment by EPashkin** [comment]
This intended that inner fields not comma-separated?
```
impl ::std::fmt::Debug for AtkObject {
    fn fmt(&self, f: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
        write!(f, "AtkObject @ {:?} {{ {:?} {:?} {:?} {:?} {:?} {:?} {:?} }}",
               self as *const _,
               self.parent,
               self.description,
               self.name,
               self.accessible_parent,
               self.role,
               self.relation_set,
               self.layer)
    }
}
```

**Comment by GuillaumeGomez** [comment]
Hum yes. Do you want comma?

**Comment by EPashkin** [comment]
Then its normal, just strange.

File: src/codegen/sys/lib_.rs (hunk 1 above)
Comment (EPashkin) [reply src/codegen/sys/lib_.rs]: Yes very strange

**Comment by GuillaumeGomez** [comment]
I'll add the comma tomorrow then. ;)

**Comment by GuillaumeGomez** [comment]
Updated.

**Comment by sdroege** [comment]
For printing things struct-like, please use the functions from Formatter like debug_struct()

File: src/codegen/sys/lib_.rs [hunk 5]
```diff
@@ -376,6 +381,57 @@ fn generate_unions(w: &mut Write, env: &Env, items: &[&Union]) -> Result<()> {
...
+                 }
+             })
+             .collect::<Vec<String>>()
+             .join(",\n");
+
+    generate_debug_impl(
+        w,
+        name,
+        &format!(
+            "write!(f, \"{name} @ {{:?}} {{{{ {fields}}}}}\",\n\
```
Comment (EPashkin) [inline src/codegen/sys/lib_.rs]: IMHO extra space needed `{{{{ {fields} }}}}`

File: src/codegen/sys/lib_.rs [hunk 6]
```diff
@@ -376,6 +381,57 @@ fn generate_unions(w: &mut Write, env: &Env, items: &[&Union]) -> Result<()> {
...
+        "impl ::std::fmt::Debug for {} {{\n\
+            \tfn fmt(&self, f: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {{\n\
+                \t\t{}\n\
+            \t}}\n\
+        }}\n",
+        name,
+        impl_content)
+}
+
+fn generate_fields_with_debug(w: &mut Write, name: &str, lines: &[String]) -> Result<()> {
```
Comment (EPashkin) [inline src/codegen/sys/lib_.rs]: Not `generate_debug_with_fields`?

**Comment by GuillaumeGomez** [comment]
Updated and improved. Thanks for the suggestion @sdroege!

File: src/codegen/sys/lib_.rs [hunk 7]
```diff
@@ -376,6 +381,47 @@ fn generate_unions(w: &mut Write, env: &Env, items: &[&Union]) -> Result<()> {
...
+fn generate_debug_with_fields(w: &mut Write, name: &str, lines: &[String]) -> Result<()> {
+    let fields =
+        lines.iter()
+             .filter_map(|field| {
+                 if field.contains("//") {
+                    None
+                 } else {
+                    let field_name = field.split(":").next().unwrap().trim().replace("pub ", "");
+                    Some(if field.replace(",", "").trim().ends_with("c_void") {
+                             format!("\t\t .field(\"{name}\", &\"c_void\")\n", name=field_name)
```
Comment (EPashkin) [inline src/codegen/sys/lib_.rs]: Is really need write constant string in debug result?

Comment (EPashkin) [reply src/codegen/sys/lib_.rs]: I meant: maybe better just skip this field?
Current results:
```rust
//gobject
impl ::std::fmt::Debug for GParamSpecString {
    fn fmt(&self, f: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
        f.debug_struct(&format!("GParamSpecString @ {:?}", self as *const _))
         .field("parent_instance", &self.parent_instance)
         .field("default_value", &self.default_value)
         .field("cset_first", &self.cset_first)
         .field("cset_nth", &self.cset_nth)
         .field("substitutor", &self.substitutor)
         .field("_truncated_record_marker", &"c_void")
         .finish()
    }
}

//gtk
impl ::std::fmt::Debug for GtkRcStyle {
    fn fmt(&self, f: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
        f.debug_struct(&format!("GtkRcStyle @ {:?}", self as *const _))
         .field("parent_instance", &self.parent_instance)
         .field("name", &self.name)
         .field("bg_pixmap_name", &self.bg_pixmap_name)
         .field("font_desc", &self.font_desc)
         .field("color_flags", &self.color_flags)
         .field("fg", &self.fg)
         .field("bg", &self.bg)
         .field("text", &self.text)
         .field("base", &self.base)
         .field("xthickness", &self.xthickness)
         .field("ythickness", &self.ythickness)
         .field("rc_properties", &self.rc_properties)
         .field("rc_style_lists", &self.rc_style_lists)
         .field("icon_factories", &self.icon_factories)
         .field("_truncated_record_marker", &"c_void")
         .finish()
    }
}

```

Comment (GuillaumeGomez) [reply src/codegen/sys/lib_.rs]: I want all fields to be there. So better having a useless "c_void" string than nothing from my point of view.

Comment (EPashkin) [reply src/codegen/sys/lib_.rs]: Ok

**Comment by EPashkin** [comment]
@sdroege, is good for you now?

File: src/codegen/sys/lib_.rs [hunk 8]
```diff
@@ -389,20 +435,32 @@ fn generate_classes_structs(w: &mut Write, env: &Env, classes: &[&Class]) -> Res
...
                 "{comment}#[repr(C)]\n{comment}pub struct {name}(c_void);\n",
                 comment = comment,
-                name = klass.c_type
+                name = klass.c_type,
+            ));
+            try!(generate_debug_impl(
+                w,
+                &klass.c_type,
+                &format!("write!(f, \"{name} @ {{:?}}\", self as *const _)",
+                         name=klass.c_type)
```
Comment (sdroege) [inline src/codegen/sys/lib_.rs]: You could use debug_tuple(name).field(the_pointer_string).build() here

File: src/codegen/sys/lib_.rs [hunk 9]
```diff
@@ -416,9 +474,15 @@ fn generate_interfaces_structs(w: &mut Write, interfaces: &[&Interface]) -> Resu
...
             w,
-            "#[repr(C)]\npub struct {}(c_void);",
+            "#[repr(C)]\npub struct {}(c_void);\n",
             interface.c_type
         ));
+        try!(generate_debug_impl(
+            w,
+            &interface.c_type,
+            &format!("write!(f, \"{name} @ {{:?}}\", self as *const _)",
+                     name=interface.c_type)
```
Comment (sdroege) [inline src/codegen/sys/lib_.rs]: And here

File: src/codegen/sys/lib_.rs [hunk 10]
```diff
@@ -456,15 +529,24 @@ fn generate_records(w: &mut Write, env: &Env, records: &[&Record]) -> Result<()>
...
                 try!(writeln!(w, "{}{}", comment, line));
             }
             try!(writeln!(w, "{}}}\n", comment));
         }
+        if lines.is_empty() || !can_generate_fields_debug(&record.fields) {
+            try!(generate_debug_impl(
+                w,
+                &record.c_type,
+                &format!("write!(f, \"{name} @ {{:?}}\", self as *const _)",
+                         name=record.c_type)
```
Comment (sdroege) [inline src/codegen/sys/lib_.rs]: And here

**Comment by sdroege** [comment]
Feel free to ignore that comment, it's all good to go for me.

**Comment by GuillaumeGomez** [comment]
To give @EPashkin a bit of context: @sdroege and I discussed about it and globally what came was:

> me: I don't see the use for the `debug_tuple`. If we have the need of it some day, I'll add it but for now it's not really useful.
> @sdroege: Ok.

**Comment by GuillaumeGomez** [comment]
I'll let you handle the merging @EPashkin.

**Comment by GuillaumeGomez** [comment]
Sorry but I want to quickly move towards the next release so I'll merge this and make a PR for the sys repository.

**Comment by EPashkin** [comment]
Thanks, @GuillaumeGomez

**Pull request merged**